### DISCUSSION TITLE
File system change events support for mounted directories

### DISCUSSION THREAD

**Discussion by balajiv113** [lima-vm/lima#1853] [Ideas]
**Problem**
For mounted directories inotify events are not fired on guest. Due to this some application that depends on inotify will not work (webpack, build tools etc)

**Approach**
We can do the following to support inotify for all types of mounted directories (sshfs, virtiofs, 9p)
- Use https://github.com/rjeczalik/notify to get native inotify events from host machine (For all mounted volumes we can bind this to get events)
- These events from host will be sent to guest via guestagent
- In guestagent, we will touch files to trigger a inotify change (This can be replaced with a native support if kernel supports injecting inotify events later)

This approach is similar to colima except for using the guestagent natively

This model should support all possible combination that we currently have,
- Support on different OS (Linux, macOS, Windows)
- Support all mount types as the events are initiated from host machine

**Comment by balajiv113** [comment]
Should we think about revamping our host <-> guest communication ?? Currently it follows a httpclient (with a underlying unix socket) model. 
Should we look into using more of like socket based communication directly to support bi-directional push of events ??

Am looking into this as we will be getting more events now than before, the event flow will be from host -> guest here. Also trying to avoid polling which might add delays as well

**Reply by balajiv113** [reply]
cc: @lima-vm/maintainers

**Reply by AkihiroSuda** [reply]
You can just use HTTP POST for posting events, but if it is problematic we can consider switching away to gRPC.

**Reply by balajiv113** [reply]
I have been facing some EOF and guest agent close on implementing inotify. Will check it out. If needed will try with grpc as well

**Reply by balajiv113** [reply]
With the draft PR am seeing EOF on high load but reconnects later. Will do one more test and see.

On the other hand, We won't be able to use grpc, as they won't provide support for using raw connection like in httpclient (it will be needed for our vsock use cases)

**Comment by abiosoft** [comment]
> In guestagent, we will touch files to trigger a inotify change (This can be replaced with a native support if kernel supports injecting inotify events later)

Just an FYI that while `touch` works for most cases, there were two main issues encountered during the implementation in Colima.

1.  it was problematic with Vim
2. it propagates back to the host (from the VM), which can cause an infinite file modification cycle, if not handled.

**Reply by balajiv113** [reply]
@abiosoft Thanks for the heads up. 
Will check it out and handle it accordingly
